### PR TITLE
DO-54 Automated data migration synchronisation

### PR DESCRIPTION
**Pull request by alvinypyim**
# Why?

We need to control data migration happening before each container starts so that there will always be one single instance running the migration; otherwise there will be race conditions.

To achieve that, I'm using Redis lock https://redis.io/topics/distlock as a centralised lock for the instances to stop other instances from starting migration.

The logic is as follows.

1. Obtain the Redis lock.
1. Halt the deployment if the previous migration failed.
1. Mark the migration failed in Redis (for extreme cases, e.g. complete network failure).
1. Run the migration task.
1. If the task failed, halt the deployment.
1. Remove the failure marker from Redis.
1. Release the lock.
1. Start the webserver.

# Testing

I've tested this repeatedly with different scenarios locally. I will perform an end-to-end test once the 3 parts are merged.

### CODE REVIEW

**Review by blaknite — APPROVED** [review]

**Comment by alvinypyim** [comment]
Please remember to update the README file regarding this new feature.

**Review by alxberardi — COMMENTED** [review]
What happens if execution fails after the lock has been granted? How can we release the lock? Should there be a reasonable timeout?

**Comment by alvinypyim** [comment]
> What happens if execution fails after the lock has been granted? How can we release the lock? Should there be a reasonable timeout?

I assume we are talking about un-handled errors.

1. If there is an exception thrown from inside `$mutex->synchronized(...)`, the lock will be released instantly.
1. if the script is terminated forcefully for whatever reasons (e.g. the instance is terminated), then the lock will still be in Redis until 3,600 seconds has passed.

Both of these cases are fatal errors; and they requires human interventions to resolve.

With that said, I think there is a case the script didn't handle nicely. ⬇️ If the lambda didn't finish in time, I don't know what `$exitCode` will be.

```
$mutex = new PHPRedisMutex([$redis], KEY_IN_PROGRESS, 3600);
$exitCode = $mutex->synchronized(
  function () use ($redis) {
    // ...
  }
);
```

**Comment by alvinypyim** [comment]
> Please remember to update the README file regarding this new feature.

✅ Done!

**Comment by alvinypyim** [comment]
> With that said, I think there is a case the script didn't handle nicely. ⬇️ If the lambda didn't finish in time, I don't know what $exitCode will be.

Addressed as well.  Now timing out and all other exceptions will be treated as migration failure; thus halting all the subsequent deployment.

**Comment by alvinypyim** [comment]
There is an edge case that I'd like to address.

If the synchronised block timed out, the lock will be released immediately before writing `migration-failed` to Redis.  The other instance might start migration before seeing `migration-failed` in Redis.

**Comment by alvinypyim** [comment]
I also addressed the potential race condition 👍.

I consider this solution quite robust.

File: README.md [hunk 1]
```diff
@@ -15,9 +15,40 @@ The only requirement is Docker.
...
+### Support zero-downtime deployment
+
+When upgrading Craft schema or plugins, i.e. data migration, the default Craft
+behaviour requires us to press a button through the admin page.  This base image
+automated that so that upgrading Craft schema and plugins is done automatically
+before starting the web server.
+
+Also, parallel deployment to multi-instances is supported.  When deploying a new
+version to multiple instances at the same time, one of the instances will
+started the data migration; all other instances will wait for the first data
```
Comment (alxberardi) [inline README.md]: engrish 😝

**Review by alxberardi — APPROVED** [review]

**Pull request merged**